cuda/solver/lower_trs_kernels.cu

```diff
@@ -62,6 +64,173 @@ namespace cuda {
 namespace lower_trs {
 
 
+constexpr int default_block_size = 512;
+
+
+template <typename ValueType>
+__device__ std::enable_if_t<std::is_floating_point<ValueType>::value, ValueType>
+load(const ValueType *values, int index)
+{
+    const volatile ValueType *val = values + index;
+    return *val;
+}
+
+template <typename ValueType>
+__device__ std::enable_if_t<std::is_floating_point<ValueType>::value,
+                            thrust::complex<ValueType>>
+load(const thrust::complex<ValueType> *values, int index)
+{
+    auto real = reinterpret_cast<const ValueType *>(values);
+    auto imag = real + 1;
+    return {load(real, 2 * index), load(imag, 2 * index)};
+}
+
+template <typename ValueType>
+__device__ void store(
+    ValueType *values, int index,
+    std::enable_if_t<std::is_floating_point<ValueType>::value, ValueType> value)
+{
+    volatile ValueType *val = values + index;
+    *val = value;
+}
+
+template <typename ValueType>
+__device__ void store(thrust::complex<ValueType> *values, int index,
```

Comment:
Do you still see performance degredation from using these separate stores/loads? Should we just remove them?

Reply:
No, at least not from quick tests. Forceinlining also doesn't change it, so I guess we're good to go with these.